File: src/web/Utils/SECTickerSyncService.cs
using System;
using System.Linq;
using System.Threading.Tasks;
using core.fs.Adapters.Logging;
using core.fs.Adapters.Storage;
using Microsoft.FSharp.Core;
using secedgar;

namespace web.Utils;

public class SECTickerSyncService
{
    private readonly IAccountStorage _accountStorage;
    private readonly EdgarClient _edgarClient;
    private readonly ILogger _logger;

    public SECTickerSyncService(
        IAccountStorage accountStorage,
        EdgarClient edgarClient,
        ILogger logger)
    {
        _accountStorage = accountStorage;
        _edgarClient = edgarClient;
        _logger = logger;
    }

    public async Task Execute()
    {
        try
        {
            _logger.LogInformation("Starting SEC ticker-to-CIK mapping sync");

            // Check when we last updated
            var lastUpdatedOpt = await _accountStorage.GetTickerCikLastUpdated();
            
            bool shouldUpdate;
            if (FSharpOption<DateTimeOffset>.get_IsSome(lastUpdatedOpt))
            {
                var lastUpdate = lastUpdatedOpt.Value;
                var daysSinceUpdate = (DateTimeOffset.UtcNow - lastUpdate).TotalDays;
                
                if (daysSinceUpdate >= 7.0)
                {
                    _logger.LogInformation($"Last ticker sync was {daysSinceUpdate:F1} days ago, updating...");
                    shouldUpdate = true;
                }
                else
                {
                    _logger.LogInformation($"Ticker data was updated {daysSinceUpdate:F1} days ago, skipping sync");
                    shouldUpdate = false;
                }
            }
            else
            {
                _logger.LogInformation("No previous ticker sync found, performing initial sync");
                shouldUpdate = true;
            }

            if (shouldUpdate)
            {
                // Fetch latest company tickers from SEC
                var companyTickers = await _edgarClient.FetchCompanyTickers();
                
                _logger.LogInformation($"Fetched {companyTickers.Count} company tickers from SEC");

                // Convert to TickerCikMapping format
                var now = DateTimeOffset.UtcNow;
                var mappings = companyTickers.Values
                    .Select(entry => new TickerCikMapping
                    {
                        Ticker = entry.ticker.ToUpperInvariant(),
                        Cik = entry.cik_str.PadLeft(10, '0'), // Pad CIK to 10 digits with leading zeros
                        Title = entry.title,
                        LastUpdated = now
                    })
                    .ToArray();

                // Save to database
                await _accountStorage.SaveTickerCikMappings(mappings);

                _logger.LogInformation($"Successfully synced {mappings.Length} ticker-to-CIK mappings");
            }
        }
        catch (Exception ex)
        {
            _logger.LogError($"Error syncing SEC ticker mappings: {ex.Message}");
            _logger.LogError(ex.ToString());
        }
    }
}

File: src/core.fs/Services/SECTickerSyncService.fs
module core.fs.Services.SECTickerSyncService

open System
open System.Linq
open System.Threading.Tasks
open core.fs.Adapters.Logging
open core.fs.Adapters.Storage
open core.fs.Adapters.SEC

type SECTickerSyncService(
    accountStorage: IAccountStorage,
    secFilingsClient: ISECFilings,
    logger: ILogger) =
    
    member _.Execute() = task {
        try
            logger.LogInformation("Starting SEC ticker-to-CIK mapping sync")
            
            // Check when we last updated
            let! lastUpdated = accountStorage.GetTickerCikLastUpdated()
            
            let shouldUpdate = 
                match lastUpdated with
                | Some lastUpdate ->
                    let daysSinceUpdate = (DateTimeOffset.UtcNow - lastUpdate).TotalDays
                    if daysSinceUpdate >= 7.0 then
                        logger.LogInformation($"Last ticker sync was {daysSinceUpdate:F1} days ago, updating...")
                        true
                    else
                        logger.LogInformation($"Ticker data was updated {daysSinceUpdate:F1} days ago, skipping sync")
                        false
                | None ->
                    logger.LogInformation("No previous ticker sync found, performing initial sync")
                    true
            
            if shouldUpdate then
                // Fetch latest company tickers from SEC
                let! companyTickers = secFilingsClient.FetchCompanyTickers()
                
                logger.LogInformation($"Fetched {companyTickers.Count} company tickers from SEC")
                
                // Convert to TickerCikMapping format
                let now = DateTimeOffset.UtcNow
                let mappings = 
                    companyTickers.Values
                    |> Seq.map (fun (entry: CompanyTickerEntry) ->
                        {
                            Ticker = entry.ticker.ToUpperInvariant()
                            Cik = entry.cik_str.PadLeft(10, '0') // Pad CIK to 10 digits with leading zeros
                            Title = entry.title
                            LastUpdated = now
                        })
                    |> Seq.toArray
                
                // Save to database
                do! accountStorage.SaveTickerCikMappings(mappings)
                
                logger.LogInformation($"Successfully synced {mappings.Length} ticker-to-CIK mappings")
            
        with ex ->
            logger.LogError($"Error syncing SEC ticker mappings: {ex.Message}")
            logger.LogError(ex.ToString())
    }
